SBOM ingestion

barv-jfrog opened the pull request:
- [ ] All [tests](https://github.com/jfrog/jfrog-client-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] All [static analysis checks](https://github.com/jfrog/jfrog-client-go/actions/workflows/analysis.yml) passed.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

The point of this PR is part of 3 PRs that enable the command "jf sbom-enrich" which takes XML/JSON of a CycloneDX SBOMs of packages/dockers, sends it to XRAY for scanning (to find vulnerabilities) and print out the input (which is xml/json cyclonedx) with vulnerabilities section that was found in XRAY. It basically uses a new API that can digest these new formats, but it still uses GraphScan and ExportGraphScan which are pre-existing APIs. That's why there are some similarities between the two.

barv-jfrog commented:
I have read the CLA Document and I hereby sign the CLA

attiasas reviewed (changes requested):
* Add description to the PR
* Change the README.md to include those changes.
* Add tests at `tests` for the new service

attiasas commented on xray/services/enrich.go:
```diff
@@ -0,0 +1,99 @@
+	utils.SetContentType("application/json", &httpClientsDetails.Headers)
+	url := ss.XrayDetails.GetUrl() + importGraph
+	if err != nil {
+		utils.SetContentType("application/xml", &httpClientsDetails.Headers)
+		url = ss.XrayDetails.GetUrl() + importGraphXML
```
please add comment why this is a fallback

barv-jfrog replied:
added

attiasas commented on xray/services/enrich.go:
```diff
@@ -0,0 +1,99 @@
+	httpClientsDetails := ss.XrayDetails.CreateHttpClientDetails()
+	var v interface{}
+	err := json.Unmarshal(scanParams.SBOMInput, &v)
+	utils.SetContentType("application/json", &httpClientsDetails.Headers)
+	url := ss.XrayDetails.GetUrl() + importGraph
```
only set if err is not nil

barv-jfrog replied:
ok

attiasas commented on xray/services/enrich.go:
```diff
@@ -0,0 +1,99 @@
+	utils.SetContentType("application/json", &httpClientsDetails.Headers)
+
+	// The scan request may take some time to complete. We expect to receive a 202 response, until the completion.
+	endPoint := ss.XrayDetails.GetUrl() + scanGraphAPI
```
is it suppost to be `scanGraphAPI`? or is it `importGraph` \ `importGraphXML` 
if scanGraphAPI, please add a comment about this.
If one of the other two, how can you make sure which endpoint you sent the request so you can access for results?

barv-jfrog replied:
scangraph api and added a comment

attiasas commented on xray/services/enrich.go:
```diff
@@ -0,0 +1,99 @@
+		Timeout:         defaultMaxWaitMinutes,
+		PollingInterval: defaultSyncSleepInterval,
+		PollingAction:   pollingAction,
+		MsgPrefix:       "Get Dependencies Scan results... ",
```
MsgPrefix is OK? or is it a copy-paste mistake?

barv-jfrog replied:
its ok

attiasas commented on xray/services/enrich.go:
```diff
@@ -0,0 +1,99 @@
+
+	// The scan request may take some time to complete. We expect to receive a 202 response, until the completion.
+	endPoint := ss.XrayDetails.GetUrl() + scanGraphAPI
+	// Modify endpoint if XSC is enabled
```
XSC is supporting this? if comment in not needed, remove please

barv-jfrog replied:
removed

attiasas reviewed (changes requested):
* Add description to the PR
* Change the README.md to include those changes.
* Add tests at `tests` for the new service

barv-jfrog commented on xray/services/enrich.go:
```diff
@@ -0,0 +1,99 @@
+	endPoint += "/" + scanId
+	endPoint += includeVulnerabilitiesParam
+	log.Info("Waiting for enrich process to complete on JFrog Xray...")
+	pollingAction := func() (shouldStop bool, responseBody []byte, err error) {
+		resp, body, _, err := ss.client.SendGet(endPoint, true, &httpClientsDetails)
+		if err != nil {
+			return true, nil, err
+		}
+		if err = errorutils.CheckResponseStatusWithBody(resp, body, http.StatusOK, http.StatusAccepted); err != nil {
+			return true, nil, err
+		}
+		// Got the full valid response.
+		if resp.StatusCode == http.StatusOK {
+			return true, body, nil
+		}
+		return false, nil, nil
+	}
```
I'm not sure its possible since it relies on parameters that are different between the two functions.

barv-jfrog replied:
fixed

barv-jfrog commented:
The point of this PR is part of 3 PRs that enable the command "jf sbom enrich" which takes XML/JSON of a CycloneDX SBOMs of packages/dockers, sends it to XRAY for scanning (to find vulnerabilities) and print out the input (which is xml/json cyclonedx) with vulnerabilities section that was found in XRAY. It basically uses a new API that can digest these new formats, but it still uses GraphScan and ExportGraphScan which are pre-existing APIs. That's why there are some similarities between the two.

barv-jfrog commented:
> * Add description to the PR
> * Change the README.md to include those changes.
> * Add tests at `tests` for the new service

Added description and changed README.md to include those changes.
Regarding tests - I use similar functions to Scan just with different XRAY Api. I have not seen tests for these functions in scan service. 
@attiasas
